index.ts

```diff
@@ -0,0 +1,187 @@
+function sum(x: string, y: string): string;
+
+function sum(x: number | string, y: number | string): number | string {
+       if (typeof x === 'number' && typeof y === 'number') {
+              return x + y;
+       } else {
+              return (x.toString() + y.toString());
+       }
+
+}
+console.log(sum(11, 3))
+console.log(sum(11, "3"))
+console.log(sum("12", "3"))
+
+
+// Затипизировать this
+function getName(this: { [id: string]: { name: string } }, id: string) {
+       return this[id]?.name;
+}
+
+
+// Затипизировать возвращаемое значение
+const sayHi = (): void => {
+       console.log('hi');
+}
+
+// Тип значение которого может быть тип Book или Car
+interface Book { }
+interface Car { }
+type BookOrCar = Book | Car;
+
+// Используя типы House, City, Country создать новый тип FullAddress который включает все свойства вышеперечисленных
+interface House { apartment: number }
+interface City { zipCode: number }
+interface Country { name: string }
+type FullAddress = House & City & Country;
+const address: FullAddress = {
+       apartment: 15,
+       zipCode: 4,
+       name: "Minsk"
+}
+
+// переменная с типом строка, начинающаяся с префикса "id:" и дальше числовое id. Например, let myId = 'id:2'
+type ID = `${"id:"}${number}`
+let prefixedId: ID = 'id:2';
+
+
+// Нам точно известно, что в качестве параметра "x" придет число. Изменять типизацию функции (параметров или возвращаемого значения) нельзя.
+const f = (x: number | string): number => {
+       // На данной строке нужно исправить текущую ошибку путем приведения типа "x" от текущего число или строка к только число
+       return +x;
+}
+
+// переменная с типом массива строк, который нельзя изменять (например, пушить в него значения)
+let readonlyStringArray: readonly string[] = ["hello"];
+
+// Сделать переменную неизменяемой (чтобы на уровне типизации ее нельзя было мутировать)
+type user = {
+       readonly id: number,
+       readonly name: string,
+}
+const USER: user = {
+       id: 1,
+       name: 'Alex',
+};
+
+// Написать реализацию функции getArea, чтобы она рассчитывала и возвращала площадь фигуры
+interface Circle { type: 'circle', radius: number }
+interface Square { type: 'square', side: number }
+const pi = Math.PI;
+
+const getArea = (value: Circle | Square): number => {
+       if (value.type === "circle") {
+              return pi * Math.pow(value.radius, 2)
+       } else {
+              return Math.pow(value.side, 2)
+       }
+}
+
+// У функции toLowerCase ошибка типизации т.к. value может быть числом, а у числа нету метода toLowerCase. 
+// Изменять типизацию функции (параметров или возвращаемого значения) нельзя.
+// Нужно решать проблему любыми двумя способами.
+const toLowerCase1 = (value: string | number): number | string => {
+       if(typeof value === "string"){
+       return value.toLowerCase();
+       } else {
+              return value;
+       }
+}
+
+const toLowerCase2 = (value: string | number): number | string => {
```

Comment:
оператор as приводит типы тайпскрипта, т.е. когда код будет скомпилирован в js оператор as будет удален ровно как и все типы. Чтобы приведение оставалось, его нужно делать средствами самого js, например
```suggestion
const toLowerCase2 = (value: string | number): number | string => String(value).toLowerCase();
```